Added gzip compression/decompression

MCJack123 opened the pull request:
I added a basic LibDeflate:DecompressGzip function. Tested with gzip 1.6 and Lua 5.1. Uses some bit functions that are included, but this may affect code style.

MCJack123 committed:
Added gzip decompression

MCJack123 committed:
Fixed issues

MCJack123 committed:
Update README.md

MCJack123 committed:
Accidentally pasted twice

SafeteeWoW commented on LibDeflate.lua:
```diff
@@ -2825,6 +2825,70 @@ do
+		error(("Usage: LibDeflate:DecompressGzip(str): "..arg_err), 2)
+	end
+	if string_byte(string.sub(str, 1, 1)) ~= 31 or string_byte(string.sub(str, 2, 2)) ~= 139 then
+		error("LibDeflate:DecompressGzip(str): Not valid gzip data", 2)
```
Should not throw when function argument type is valid, but data cant be decompressed.
return nil, error_code in this case.
For example,
return nil, 1

SafeteeWoW commented on LibDeflate.lua:
```diff
@@ -2825,6 +2825,70 @@ do
+        offset = offset + 4
+    end
+	return DecompressDeflateInternal(string.sub(str, offset + 1, -8))
+end
```
The input data must be validated to comply RFC 1952.
You need to implement CRC32 and check the decompress result against the CRC32 in the input stream.
See RFC1952 Section  2.3.1.2.

SafeteeWoW commented:
Hi MCJack123. I have reviewed your code. Thank you for your contribution.

It is okay to introduce dependencies, as long as that is only needed for gzip functionality.
Can you implement some of the following checklists when you have time?
I currently have a busy work schedule, will help in my early May vocation.

- [ ] Pass CI. [tests/Test.lua:2954](https://github.com/SafeteeWoW/LibDeflate/blob/07f9d36b9466f25f721b46a2c815c1d0294d93c5/tests/Test.lua#L2954). Add your the name of the new function in the table there. This is to avoid breaking backward compatibility accidentally.

- [ ] LibDeflate only throws error when the type of input argument is invalid.  No error is thrown if argument type is valid, but data is invalid and cant be decompressed. In this case, two element, (nil, error_code) should be returned. Because I haven't defined an error code in the documentation, you can use any number as the error code in your pull request.

- [ ] Make gzip decompressor comply RFC 1952 standard. The integrity of input data to be decompressed must be checked.

- [ ] > RFC 1952 2.3.1.2. Compliance
         A compliant decompressor must check ID1, ID2, and CM, and
         provide an error indication if any of these have incorrect
         values.  It must examine FEXTRA/XLEN, FNAME, FCOMMENT and FHCRC
         at least so it can skip over the optional fields if they are
         present.  It need not examine any other part of the header or
         trailer; in particular, a decompressor may ignore FTEXT and OS
         and always produce binary output, and still be compliant.  A
         compliant decompressor must give an error indication if any
         reserved bit is non-zero, since such a bit could indicate the
         presence of a new field that would cause subsequent data to be
         interpreted incorrectly.

- [ ] Support gzip compression and comply RFC 1952 standard.
> RFC 1952 2.3.1.2. Compliance
         A compliant compressor must produce files with correct ID1,
         ID2, CM, CRC32, and ISIZE, but may set all the other fields in
         the fixed-length part of the header to default values (255 for
         OS, 0 for all others).  The compressor must set all reserved
         bits to zero.

- [ ] Support gzip compression and decompression when calling LibDeflate from the commandline.

- [ ] Add tests for your code, and maintain 100.00% code CI tests coverage.
Read [tests/README.md](https://github.com/SafeteeWoW/LibDeflate/blob/master/tests/README.md) and [.appveyor.yml](https://github.com/SafeteeWoW/LibDeflate/blob/master/.appveyor.yml) for tests instructions.

Thanks for your contributions.

MCJack123 committed:
Fixed compliance, added CRC checks

MCJack123 committed:
Added compressor

MCJack123 committed:
Added functions to test

MCJack123 committed:
Added command-line (+ComputerCraft support)

MCJack123 committed:
Added CRC32 to tests

MCJack123 committed:
Fixed local leak

MCJack123 committed:
Fixed CC compatibility for large files

MCJack123 commented:
I've added a compressor and fixed compliance, but something is causing an error in the tests, so I'm going to need to spend a bit more time debugging that.

MCJack123 committed:
Update README.md

SafeteeWoW commented:
As long as most texts passed, it's okay. I'll fix failed tests.   You can search "Fail" in the CI log to figure out which testbis failing. Some tests I wrote are probably not needed.

SafeteeWoW commented:
Dont assume "os" and "io" module exists. LibDeflate is originally written for games embeded Lua, which strip away os and io module

SafeteeWoW commented:
Can you give an instruction how to work with ComputerCraft,so I can test it?

SafeteeWoW commented:
What is the size of your big files? LibDeflate  is slow when work with large files. It should work for large files though. I did test  50M input data. But anyway, prefer compressor written in C with Lua binding when working on big files.

SafeteeWoW commented on LibDeflate.lua:
```diff
@@ -2762,7 +2766,12 @@ local function Inflate(state)
-		end
+        end
+        -- ComputerCraft requires this for long-running processes
+        if os and os.pullEvent then
```
Don't think this is needed.
If the input string is too long. Caller can split the string into multiple parts and call LibDeflate:Compress multiple times and concat the input string.

SafeteeWoW commented on LibDeflate.lua:
```diff
@@ -2188,6 +2188,43 @@ function LibDeflate:CompressZlibWithDict(str, dictionary, configs)
+        -- ComputerCraft's os.time() gives in-game time, os.epoch gives POSIX time in ms
+    elseif os.time() < 30 then return 0 -- ComputerCraft 1.79 and below don't have os.epoch(), so no time.
+    else return os.time() end -- All other Luas.
+end
```
It's possible that module "os" does not exist. return 0 in that cae.

SafeteeWoW commented on LibDeflate.lua:
```diff
@@ -3004,6 +3040,7 @@ function LibDeflate:DecompressGzip(str)
     src_checksum = bnot(src_checksum)
     local target_checksum = self:CRC32(res)
+    print(src_checksum, target_checksum)
     if xor(src_checksum, target_checksum) ~= 0xFFFFFFFF then return nil, -2 end
```
"print" function shouldn't be here.

MCJack123 replied:
Removed in latest commit.

SafeteeWoW commented on LibDeflate.lua:
```diff
@@ -3332,19 +3560,54 @@ LibDeflate.internals = {
+local arg = _G.arg
+local debug = debug
+if shell then 
+    arg = {...} 
```
Is "shell" a global variable? If yes, use "_G.shell" to access it.
I prefer to use another way to check if ComputerCraft shell exists.
Check _G.shell seems not safe.

MCJack123 replied:
Shell is an environmental variable, so I can't use _G.shell to check it.

MCJack123 replied:
Also, shell is only available when called from the shell.

SafeteeWoW replied:
The more standard way to fetch environment variable is to use "os.getenv" method.
Does ComputerCraft create "shell" variable as a local variable?
If that is the case, use "debug.getlocal" to get the value of "shell" variable.

Don't access non-standard global variable like this. Also,

MCJack123 replied:
ComputerCraft has neither os.getenv or the debug library, but it does have an _ENV upvalue as in Lua 5.2+, as well as getfenv() as in Lua 5.1. Neither of these work in all Lua versions, though, so I'm not sure which one I should check for, since both ways will break for at least one version. (`_ENV.shell == getfenv(1).shell` = true)

SafeteeWoW replied:
Plan to remove direct support for ComputerCraft specific APIs in the code of LibDeflate.

SafeteeWoW commented on LibDeflate.lua:
```diff
@@ -3332,19 +3560,54 @@ LibDeflate.internals = {
+    end}
+    os.exit = function() error() end
+    io.stderr = {write = function(self, text) printError(text) end}
+end
```
Don't modify anything in other global table, unless it's a module manager.

SafeteeWoW commented:
LibDeflate gzip functionality should be tested against "gzip"
See [tests/Test.lua:444](https://github.com/SafeteeWoW/LibDeflate/blob/07f9d36b9466f25f721b46a2c815c1d0294d93c5/tests/Test.lua#L444)

SafeteeWoW commented on LibDeflate.lua:
```diff
@@ -1860,13 +1991,12 @@ local function Deflate(configs, WriteBits, WriteString, FlushWriter, str
 					if tSize == 1 then
-						hash_tables[k] = nil
+						--hash_tables[k] = nil -- seems to be causing problems
 					else
```
Don't think this causes the problem

MCJack123 replied:
I was getting an error about an invalid key to next() because the key was being deleted, and then accessed in the for loop. I usually fix this problem by making a new table of keys to set to nil, which I then loop over after accessing the first table.
    
    local tab, keys_to_delete = {}
    for k,v in pairs(tab) do
        if condition then
            table.insert(keys_to_delete, k)
        end
    end
    for _,k in pairs(keys_to_delete) do tab[k] = nil end

Something like that.

SafeteeWoW replied:
This does not reaccess the table element after setting it to nil. Can you reproduce it in a standalone Lua interpreter?  I believe ComputerCraft uses LiaJit 2.0.3.  Can you reproduce it in Lua 2.0.3 outside ComputerCraft environment?

SafeteeWoW replied:
> -- Sample code to remove keys whose value is 0
for k, v in t do
    if v == 0 then
        k[v] = nil
    end
end

SafeteeWoW replied:
The code above is the standard way to conditionally remove the element in the table.
I suppose this is a ComputerCraft specific error? 
If that is the case, I could use your workaround by "by making a new table of keys to set to nil, which I then loop over after accessing the first table", but for the performance issue, this is only activated when ComputerCraft is detected. The origin code should be used otherwise.

SafeteeWoW commented on LibDeflate.lua:
```diff
@@ -1881,7 +2011,12 @@ local function Deflate(configs, WriteBits, WriteString, FlushWriter, str
 					end
```
I probably should check "newSize" here
`
if newSize > 0 then
    hash_tables[k] = new
else
    hash_tables[k] = {}
end
`

SafeteeWoW replied:
Also the code style needs to be fixed here, I use "under_score" for variables. CamelCase for functions.

SafeteeWoW commented:
I will work on your PR on May 1st, hopefully get a Beta version on May 2nd.

MCJack123 commented:
I've added a compressor and fixed compliance, but something is causing an error in the tests, so I'm going to need to spend a bit more time debugging that.

> 
> 
> What is the size of your big files? LibDeflate is slow when work with large files. It should work for large files though. I did test 50M input data. But anyway, prefer compressor written in C with Lua binding when working on big files.

I need to be able to compress 700k+ files in ComputerCraft, which is pure Lua. ComputerCraft will halt the Lua computer if it runs for 7 seconds without calling `os.pullEvent()`. This is to prevent infinite loops in programs, but can cause problems for programs that require long processing.

MCJack123 committed:
Fixed global modification

MCJack123 committed:
Fixed time/os

MCJack123 committed:
Added (De)compressGzip to tests

MCJack123 committed:
Update install-compressor-decompressor.cmd

MCJack123 committed:
Fixed math.log

MCJack123 committed:
...and math.floor

MCJack123 committed:
...and xor

MCJack123 committed:
Update LibDeflate.lua

MCJack123 committed:
Fixed block type

MCJack123 committed:
Checking CRCs of gzips

MCJack123 committed:
Checking gzip CRC

MCJack123 committed:
Fix

MCJack123 committed:
Fixed bnot implementation

MCJack123 committed:
Fixed DecompressGzip unprocessed bytes

MCJack123 committed:
Added CRC-32 test

SafeteeWoW commented:
Compress file with 700k size, even on compression level 9, shouldn't take 7s.
If long processing causes the problem, you can split the file data into multiple parts and call LibDeflate:Compress multiple times and then concat the result

SafeteeWoW commented:
btw, your code is failing the static analyzer check in CI. See CI log at the top.

SafeteeWoW commented:
I have pushed several changes into the **develop** branch of this repository.

MCJack123 committed:
Fixed some errors

SafeteeWoW commented on LibDeflate.lua:
```diff
@@ -1992,11 +1998,13 @@ local function Deflate(configs, WriteBits, WriteString, FlushWriter, str
+                    if tSize == 1 then
+                        if _G.os and _G.os.pullEvent then table.insert(delete, k)
+						else hash_tables[k] = nil end
 					else
```
Can you confirm it is a ComputerCraft specific workaround?

MCJack123 replied:
`os.pullEvent` always exists on ComputerCraft, and does not seem to exist on anything else.

SafeteeWoW replied:
Not going to merge changes which includes ComputerCraft specific API.
Caller should be able to workaround "os.pullEvent" by splitting big input into multiple parts and compress multiple times.

SafeteeWoW commented on LibDeflate.lua:
```diff
@@ -312,6 +312,143 @@ do
+    bnot = bit.bnot
+    xor = bit.bxor
+    lshift = bit.lshift
+    rshift = bit.rshift
```
I prefer not use define lshift and rshift, because they can be converted to mudulos and divisions and constant table lookup. See "_pow2" table at the top of LibDeflate.lua.

SafeteeWoW commented on tests/Test.lua:
```diff
@@ -346,14 +346,38 @@ end
+        end
+        if band(string.byte(compressed_data, 4), 2) == 2 then
+            offset = offset + 2
+        end
```
No need to use band here. Can be converted to modulos and division.

SafeteeWoW commented:
I will install minecraft and  computercraft to investigate computercraft specific issues

SafeteeWoW commented:
I made a mistake. Computercraft uses LuaJ as the interpreter. Not Luajit. CI will be added to LuaJ and the ComputerCraft Fork of LuaJ.

SafeteeWoW commented:
1. I will test more Lua interpreter implementations and try to add workarounds if possible.
There is a bug in LuaJ 2.0.3 that modifying table during traversal is not possible. I will add work around for it.

2. I decide to not merge code that contains ComputerCraft specific APIs because I have never played Minecraft before and I cannot test those features in CI. It's possible that ComputerCraft users write those code outside of LibDeflate without modifying LibDeflate.lua.

3. My current plans is to add tests for another 5 Lua interpreters in CI, when passed, start to merge your code.

MCJack123 committed:
Removed bitwise (besides XOR), added GetGzipInfo

MCJack123 committed:
Fixed CRC math

MCJack123 committed:
Fixed CompressGzip

MCJack123 committed:
Fixed CRC math

MCJack123 committed:
Fixed naming

MCJack123 committed:
Fixed DecompressGzip

MCJack123 commented:
I added `gzip -t -v` to the gzip tests to print info about the compressed gzip. Should I keep this for debugging, or remove it?

MCJack123 committed:
Fixed math.pow

MCJack123 committed:
Fixed strategy test

MCJack123 committed:
Actually fixed strategy tests

SafeteeWoW commented:
`gzip -t -v` should be okay, unless it makes the test log bigger than the limit of TravisCi or Appveyor.

SafeteeWoW commented:
You can convert `math.floor(x)` to `(x-x%1)`, then LibDeflate will no longer require a Lua interpreter which implements the "math" module.

SafeteeWoW commented:
Have you pass all tests locally? I suggest to pass tests locally first in LuaJIT.

SafeteeWoW commented:
CI is at least 5 times slower than the current develop branch. I guess it is caused by the CRC32. The current dev branch runs CI in 13min. This PR has exceed 50min limitation of Travis CI.

One solution is to disable gzip tests in original Lua, only enable it in LuaJIT.

SafeteeWoW commented:
I am trying to found if there is any room to optimize CRC32 without installing any "bit" module

SafeteeWoW commented:
Made a faster Pure Lua implementation of CRC32, without using any external library.
It is faster than level 1 LibDeflate compression with the same input size.

There is still room to optimize. But now I think it's safe to remove optional "bit" or "bit32"
dependencies, because it only gives 2 times speedup, for crc32 only.

It takes 200ms to generate lookup table though in classic Lua5.1 interpreter.
NOTE: In LibDeflate, lookup table should only be generated when crc32 is called for the first time.
Shouldn't generate it when LibDeflate is imported.
```
local LibDeflate = {}

local t = os.clock()
-- Xor which can only support 8bit number (0<= a,b <= 255)
-- Both arguments are uint8
local xor8 = function(a, b)
	local ret = 0
	local fact = 128
	while fact > a and fact > b do
		fact = fact / 2
	end
	while fact >= 1 do
        ret = ret + (((a >= fact or b >= fact) and (a < fact or b < fact)) and fact or 0)
        a = a - ((a >= fact) and fact or 0)
        b = b - ((b >= fact) and fact or 0)
	    fact = fact/2
	end
	return ret
end

-- table to cache the result of xor(x, y)  (0<=x,y<=255)
local xor_table = {}

for i = 0, 255 do
	local t = {}
	xor_table[i] = t
	for j = 0, 255 do
		local c = xor8(i, j)
		t[j] = c
	end
end

-- xor function which supports 32bit number
-- xor cache for 8bit number is required
local function xor32(x, y)
	local x_byte = x % 256
	local y_byte = y % 256
	local result = xor_table[x_byte][y_byte]
	x = (x-x_byte) / 256
	y = (y-y_byte) / 256

	x_byte = x % 256
	y_byte = y % 256
	result = result + xor_table[x_byte][y_byte] * 256
	x = (x-x_byte) / 256
	y = (y-y_byte) / 256

	x_byte = x % 256
	y_byte = y % 256
	result = result + xor_table[x_byte][y_byte] * 65536
	x = (x-x_byte) / 256
	y = (y-y_byte) / 256

	result = result + xor_table[x][y] * 16777216

	return result
end

-- First number is uint32, 2nd one is uint16
local function xor32_16(x, y)
	local x_byte = x % 256
	local y_byte = y % 256
	local result = xor_table[x_byte][y_byte]
	x = (x-x_byte) / 256
	y = (y-y_byte) / 256

	x_byte = x % 256
	result = result + xor_table[x_byte][y] * 256
	x = (x-x_byte) / 256

	x_byte = x % 256
	result = result + xor_table[x_byte][0] * 65536
	x = (x-x_byte) / 256

	result = result + xor_table[x][0] * 16777216

	return result
end

-- Both numbers are uint16
local function xor16(x, y)
	local x_byte = x % 256
	local y_byte = y % 256
	local result = xor_table[x_byte][y_byte]
	x = (x-x_byte) / 256
	y = (y-y_byte) / 256

	x_byte = x % 256
	y_byte = y % 256
	result = result + xor_table[x_byte][y_byte] * 256

	return result
end

-- 64K CRC32 Lookup table, unlike the classific 256 element lookup table.
local crc_table = {}
for n = 0, 65535 do
	local c = n
	for k = 0, 15 do
		local m = c % 2
		local d = (c-m)/2
		if m > 0 then
			c = xor32(0xedb88320, d)
		else
			c = d
		end
	end
	crc_table[n] = c
end

print("time took to generate lookup table: ", os.clock() - t)
local string_byte = string.byte

--- Calculate the CRC-32 checksum of the string.
-- @param s [string] the input string to calculate its CRC-32 checksum.
-- @return [integer] The CRC-32 checksum, which is greater or equal to 0,
-- and less than 2^32 (4294967296).
function LibDeflate:CRC32(s, crc)
  crc = xor32(crc or 0, 0xFFFFFFFF)
  for i=1,#s,2 do
  	local byte0, byte1 = string_byte(s, i, i+1)
  	local word = byte0 + (byte1 or 0) * 256
  	crc = xor32_16(crc_table[xor16(crc, word)], (crc-crc%65536)/65536)
  end
  return xor32(crc, 0xFFFFFFFF)
end

s=""
t={}
local table_insert = table.insert
for i=1,10000 do
	table_insert(t, "12313213213jojelasdfj;askljfldasgfjd;sajf")
end

s = table.concat(t)

print("string len: ", #s)
t = os.clock()

crc32 = LibDeflate:CRC32(s, 0)

print("crc32: "..crc32)

print("time took for crc32: ", os.clock() - t)

print("---------------------------------------------")


local bit = require("bit")
local bxor = bit.bxor
function LibDeflate:CRC32WithExternalLib(s, crc)
  crc = xor32(crc or 0, 0xFFFFFFFF)
  for i=1,#s,2 do
  	local byte0, byte1 = string_byte(s, i, i+1)
  	local word = byte0 + (byte1 or 0) * 256
  	crc = bxor(crc_table[bxor(crc, word) % 65536], (crc-crc%65536)/65536) % 4294967296
  end
  return bxor(crc, 0xFFFFFFFF) % 4294967296
end

t = os.clock()

crc32 = LibDeflate:CRC32WithExternalLib(s, 0)

print("crc32(with external library): "..crc32)

print("time took for crc32 with external library: ", os.clock() - t)
```

SafeteeWoW commented:
This project uses "TAB" as the indention, not "4 spaces". This is to follow Blizzard style, because this project is originally written as the addon to World of Warcraft.

MCJack123 committed:
Optimized CRC routine

MCJack123 committed:
Fixed remaining xor() calls

MCJack123 committed:
Fixed luacheck warnings

MCJack123 committed:
Improved CRC

SafeteeWoW commented:
The Lua used in ComputerCraft, LuaJ 2.0.3 is more than10 times slower than classic Lua interpreter. I suggest you to use LibDeflate to work than files less than 200k. Compression for 100k files is already over 10s.

SafeteeWoW commented:
If you want to increase code coverage:

Because Lua tests with code coverage tracking enabled is very slow,
code coverage is only enabled with LuaJIT.
And only subset of tests which can improve code coverage are enabled in CodeCoverage.

See AddToCoverageTest() function in tests/Test.lua

SafeteeWoW commented:
https://gist.github.com/SafeteeWoW/080e784e5ebfda42cad486c58e6d26e4

My version of crc32 updated.  Switch back to 256 elements crc32 table to solve long time of caching.

Gonna start to merge your code now.
Currently working on "test_more_interpreter" branch of this repo and will merge your code to there.

SafeteeWoW commented:
Changed my mind. The following change which workaround the slowness of ComputerCraft will be merged.

```
if _G.os and _G.os.pullEvent then -- ComputerCraft requires this for long-running processes
	os.queueEvent("nosleep")
	os.pullEvent()
end
```

SafeteeWoW commented:
But I am not sure if os.pullEvent for every block solves your problem.
The current block size is 64K, but my current experiment on LuaJ 2.0.3 shows that compression 64K files will likely exceed 7 second limit.

SafeteeWoW commented:
Your changes have been merge in test_more_interpreter.

However, all codes to work around ComputerCraft / LuaJ 2.0.3 have been removed.
You are welcome to maintain a fork of LibDeflate to make it work for ComputerCraft.

This is due to the following reasons:
1. LibDeflate running on LuaJ 2.0.3 is more than 10 times slower than classic Lua 5.1.5 interpreter,  slowing down CI and local testing process.
2. It's Lua table implementation is buggy. Don't want to fix it when most Lua interpreter implementations  I tried these days has correct Lua table implementation
3. I have never played Minecraft/ComputerCraft before. I cannot ensure the workarounds for ComputerCraft really make LibDeflate work on ComputerCraft.

MCJack123 committed:
Added new CRC, removed some CC references

MCJack123 committed:
Removed remaining CC references

MCJack123 committed:
Re-added xor32()

MCJack123 committed:
Fixed xor32

SafeteeWoW commented:
Making some relative big changes with code style fixes on test_more_interpreter branch

SafeteeWoW commented:
I am considering to add a streaming interface, which is similar to Zlib. This will allow you to compress big files with multiple separate function calls to LibDeflate with an arbitrary size of small chunk of input data, so you can os.pullEvent() or give control to other program between these function calls. Decompression will also provide similar streaming interface.

It basically looks like this:
```
state = DeflateInit(configs)
Deflate(input_chunk0, state)
-- Do whatever you want between these function calls.
Deflate(input_chunk1, state)
Deflate(input_last_chunk, state)
compressed_data = DeflateEnd(state)
```

SafeteeWoW commented:
Close for now. May add gzip function later
